### PR TITLE
Enable 0-base using config

### PR DESCRIPTION
**Pull request by RRathna**
Enables 0-base oesign configs.

### CODE REVIEW

**Review by mikbras — COMMENTED** [review]
Thank you Rathna. This looks great. A few minor comments.

And yes, you are right. We need some kind of test, but I would wait for my zero-based integration PR, because we can then check the values in the app.

Please see this PR:
https://github.com/deislabs/mystikos/pull/676

File: tools/myst/config.c [hunk 1]
```diff
@@ -476,6 +528,15 @@ int write_oe_config_fd(int fd, config_parsed_data_t* parsed_data)
 
     fprintf(out_file, "SecurityVersion=%d\n", parsed_data->oe_security_version);
 
+    fprintf(
+        out_file,
+        "CreateZeroBaseEnclave=%d\n",
+        parsed_data->oe_create_zero_base);
+    if (parsed_data->oe_create_zero_base)
+    {
+        fprintf(out_file, "StartAddress=%d\n", parsed_data->oe_start_address);
```
[the remark this answers is not in the document]
Comment (mingweishih) [reply tools/myst/config.c:537]: The default interaction between OE and Intel SGX libraries is OE passes addr as NULL, and allows the latter the choose a start address. OE currently does not allow developer to specify the start address. Do you think it's worth supporting this option?

File: tools/myst/config.c [hunk 2]
```diff
@@ -69,6 +69,32 @@ static json_result_t _extract_mem_size(
...
+    if (start_address == NULL)
+        CONFIG_RAISE(JSON_FAILED);
+
+    if (type == JSON_TYPE_INTEGER && un->integer > 0)
+        value = (uint64_t)un->integer; // number in bytes
+    else
+        CONFIG_RAISE(JSON_TYPE_MISMATCH);
+
+    if (value && value % OE_PAGE_SIZE != 0) // check if aligned to OE_PAGE_SIZE
+        ret = JSON_REASON_VALUE;
```
Comment (mingweishih) [inline tools/myst/config.c:89]: Do CONFIG_RAISE

File: tools/myst/config.c [hunk 3]
```diff
@@ -69,6 +69,32 @@ static json_result_t _extract_mem_size(
...
+    const json_union_t* un,
+    uint64_t* start_address)
+{
+    json_result_t ret = JSON_FAILED;
+    uint64_t value;
+
+    if (start_address == NULL)
+        CONFIG_RAISE(JSON_FAILED);
+
+    if (type == JSON_TYPE_INTEGER && un->integer > 0)
```
Comment (mingweishih) [inline tools/myst/config.c:83]: The greater than zero check is redundant if check in line 84 is done correctly.

File: tools/myst/config.c [hunk 4]
```diff
@@ -69,6 +69,32 @@ static json_result_t _extract_mem_size(
...
+        value = (uint64_t)un->integer; // number in bytes
+    else
+        CONFIG_RAISE(JSON_TYPE_MISMATCH);
+
+    if (value && value % OE_PAGE_SIZE != 0) // check if aligned to OE_PAGE_SIZE
```
Comment (mingweishih) [inline tools/myst/config.c:88]: Shouldn't this be if (!value || value % OE_PAGE_SIZE)?

File: tools/myst/config.c [hunk 5]
```diff
@@ -69,6 +69,32 @@ static json_result_t _extract_mem_size(
...
+        ret = JSON_REASON_VALUE;
+    else
```
Comment (mingweishih) [inline tools/myst/config.c:90]: Nit: else branch is not needed. Can just set start_address after the check.

File: tools/myst/config.h [hunk 6]
```diff
@@ -6,6 +6,7 @@
 
 #include <myst/json.h>
 #include <myst/kernel.h>
+#include <openenclave/bits/defs.h>
```
Comment (mingweishih) [inline tools/myst/config.h:9]: Is this needed?

Comment (RRathna) [reply tools/myst/config.h:9]: yes, for OE_PAGE_SIZE used in the start_address value check.

Comment (mingweishih) [reply tools/myst/config.h:9]: I think mystikos has its own PAGE_SIZE definition. Can you check?

File: tools/myst/config.c (hunk 3 above)
Comment (RRathna) [reply tools/myst/config.c:83]: Negative numbers should not be allowed as input and should be checked before the conversion to unsigned in line 84. At line 88, I am checking only for page alignment.

File: tools/myst/config.h (hunk 6 above)
Comment (RRathna) [reply tools/myst/config.h:9]: I see PAGE_SIZE = OE_PAGE_SIZE = 0x1000. Changed code to use PAGE_SIZE instead.

**Review by mikbras — APPROVED** [review]
Looks good Rathna. Thank you so much for doing this.

**Review by mingweishih — APPROVED** [review]

**Review by mingweishih — COMMENTED** [review]
Give Mike's 0-base PR has been merged already, can we also have tests for this PR?

**Comment by RRathna** [comment]
discarding this PR since pipeline is not getting triggered. New PR - https://github.com/deislabs/mystikos/pull/746
